File: api/routers/public_stats.py
from __future__ import annotations

import time
from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel
from sqlalchemy import and_, column, func, inspect, select, table
from sqlalchemy.sql.elements import quoted_name

from api.db import engine


router = APIRouter(prefix="/public", tags=["public-stats"])

_STARTED_AT = time.time()
_CACHE_TTL_SECONDS = 10
_stats_cache: dict[str, tuple[float, PublicStats]] = {}


class PublicStats(BaseModel):
    nodes: int
    shards: int
    proofs: int
    sbts_issued: int
    uptime: str
    uptime_seconds: int
    copies: int = 0


def _quoted_table(table_name: str):
    return table(quoted_name(table_name, quote=True))


async def _count_query(conn: Any, stmt: Any) -> int:
    result = await conn.execute(stmt)
    value = result.scalar()
    return int(value or 0)


async def _table_exists(conn: Any, table_name: str) -> bool:
    def _has_table(sync_conn: Any) -> bool:
        schema = None if sync_conn.dialect.name == "sqlite" else "public"
        return bool(inspect(sync_conn).has_table(table_name, schema=schema))

    return bool(await conn.run_sync(_has_table))


async def _column_exists(conn: Any, table_name: str, column_name: str) -> bool:
    if not await _table_exists(conn, table_name):
        return False

    def _has_column(sync_conn: Any) -> bool:
        schema = None if sync_conn.dialect.name == "sqlite" else "public"
        columns = inspect(sync_conn).get_columns(table_name, schema=schema)
        return any(col["name"] == column_name for col in columns)

    return bool(await conn.run_sync(_has_column))


async def _count_table_if_exists(conn: Any, table_name: str) -> int:
    if not await _table_exists(conn, table_name):
        return 0

    stmt = select(func.count()).select_from(_quoted_table(table_name))
    return await _count_query(conn, stmt)


async def _count_node_operators(conn: Any) -> int:
    if not await _column_exists(conn, "operators", "role"):
        return 0

    filters = [column("role") == "node_operator"]
    if await _column_exists(conn, "operators", "revoked_at"):
        filters.append(column("revoked_at").is_(None))

    stmt = select(func.count()).select_from(_quoted_table("operators")).where(and_(*filters))
    return await _count_query(conn, stmt)


async def _count_issued_sbts(conn: Any) -> int:
    if not await _table_exists(conn, "key_credentials"):
        return 0

    filters = []
    if await _column_exists(conn, "key_credentials", "revoked_at"):
        filters.append(column("revoked_at").is_(None))
    if await _column_exists(conn, "key_credentials", "sbt_nontransferable"):
        filters.append(column("sbt_nontransferable").is_(True))

    stmt: Any = select(func.count()).select_from(_quoted_table("key_credentials"))
    if filters:
        stmt = stmt.where(and_(*filters))
    return await _count_query(conn, stmt)


async def _distinct_column_values_if_exists(
    conn: Any, table_name: str, column_name: str
) -> set[str]:
    if not await _column_exists(conn, table_name, column_name):
        return set()

    stmt: Any = select(column(column_name)).select_from(_quoted_table(table_name)).distinct()
    result = await conn.execute(stmt)
    return {str(value) for value in result.scalars().all() if value is not None}


async def _count_non_empty_column_if_exists(conn: Any, table_name: str, column_name: str) -> int:
    if not await _column_exists(conn, table_name, column_name):
        return 0

    target: Any = column(column_name)
    stmt = (
        select(func.count())
        .select_from(_quoted_table(table_name))
        .where(and_(target.is_not(None), target != ""))
    )
    return await _count_query(conn, stmt)


def _format_uptime(seconds: int) -> str:
    if seconds < 60:
        return f"{seconds}s"
    if seconds < 3600:
        return f"{seconds // 60}m"
    if seconds < 86400:
        return f"{seconds // 3600}h"
    return f"{seconds // 86400}d"


@router.get("/stats", response_model=PublicStats)
async def get_public_stats() -> PublicStats:
    """Return aggregated public ledger statistics from an async GET endpoint.

    Args:
        None.

    Returns:
        PublicStats: Counts for nodes, shards, proofs, and process uptime.
    """
    now = time.time()
    cached = _stats_cache.get("latest")
    if cached is not None:
        cached_at, stats = cached
        if now - cached_at < _CACHE_TTL_SECONDS:
            return stats

    async with engine.connect() as conn:
        # Public stats should describe system-facing ledger entities, not
        # internal SMT rows. Counting internal tree tables makes a small local
        # database look like it has thousands of public records.
        witness_origins = await _distinct_column_values_if_exists(
            conn, "witness_observations", "origin"
        )
        nodes = await _count_node_operators(conn) + len(witness_origins)
        sbts_issued = await _count_issued_sbts(conn)

        shard_ids: set[str] = set()
        for table_name in (
            "doc_commits",
            "dataset_artifacts",
            "dataset_lineage_events",
        ):
            shard_ids.update(await _distinct_column_values_if_exists(conn, table_name, "shard_id"))
        shards = len(shard_ids)

        proofs = 0
        for table_name, column_name in (
            ("doc_commits", "zk_proof"),
            ("dataset_artifacts", "zk_proof"),
            ("credential_ledger_events", "inclusion_proof"),
        ):
            proofs += await _count_non_empty_column_if_exists(conn, table_name, column_name)

    uptime_seconds = int(now - _STARTED_AT)

    stats = PublicStats(
        nodes=nodes,
        shards=shards,
        proofs=proofs,
        sbts_issued=sbts_issued,
        uptime=_format_uptime(uptime_seconds),
        uptime_seconds=uptime_seconds,
        copies=nodes,
    )
    _stats_cache["latest"] = (now, stats)
    return stats

File: src-tauri/src/routes/public_stats.rs
use axum::{extract::State, http::StatusCode, response::Json};
use serde::Serialize;
use sqlx::PgPool;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use crate::state::{AppState, Cached};

const CACHE_TTL: Duration = Duration::from_secs(10);

#[derive(Serialize, Clone)]
pub struct PublicStats {
    pub nodes: i64,
    pub shards: i64,
    pub proofs: i64,
    pub sbts_issued: i64,
    pub uptime: String,
    pub uptime_seconds: i64,
    pub copies: i64,
}

pub async fn get_public_stats(
    State(state): State<AppState>,
) -> Result<Json<PublicStats>, StatusCode> {
    {
        let cache = state.stats_cache.lock().await;
        if let Some(c) = &*cache {
            if c.stored_at.elapsed() < CACHE_TTL {
                return Ok(Json(c.value.clone()));
            }
        }
    }

    let pool = state.pool.as_ref().ok_or(StatusCode::SERVICE_UNAVAILABLE)?;

    let (nodes, sbts_issued, shards, proofs) = tokio::try_join!(
        async { Ok::<_, StatusCode>(count_nodes(pool).await) },
        async { Ok::<_, StatusCode>(count_issued_sbts(pool).await) },
        async { Ok::<_, StatusCode>(count_distinct_shards(pool).await) },
        async { Ok::<_, StatusCode>(count_public_proofs(pool).await) },
    )?;

    let now_unix = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let uptime_seconds = (now_unix - state.started_unix).max(0);

    let stats = PublicStats {
        nodes,
        shards,
        proofs,
        sbts_issued,
        uptime: format_uptime(uptime_seconds as u64),
        uptime_seconds,
        copies: nodes,
    };

    *state.stats_cache.lock().await = Some(Cached {
        value: stats.clone(),
        stored_at: std::time::Instant::now(),
    });
    Ok(Json(stats))
}

// ─── helpers ─────────────────────────────────────────────────────────────────

async fn column_exists(pool: &PgPool, table: &str, col: &str) -> bool {
    sqlx::query_scalar::<_, bool>(
        "SELECT EXISTS(SELECT 1 FROM information_schema.columns \
         WHERE table_schema = 'public' AND table_name = $1 AND column_name = $2)",
    )
    .bind(table)
    .bind(col)
    .fetch_one(pool)
    .await
    .unwrap_or(false)
}

async fn count_nodes(pool: &PgPool) -> i64 {
    let node_operators = count_node_operators(pool).await;
    let witness_origins = count_witness_origins(pool).await;
    node_operators + witness_origins
}

async fn count_node_operators(pool: &PgPool) -> i64 {
    if !column_exists(pool, "operators", "role").await {
        return 0;
    }
    if column_exists(pool, "operators", "revoked_at").await {
        sqlx::query_scalar::<_, i64>(
            "SELECT COUNT(*) FROM operators WHERE role = 'node_operator' AND revoked_at IS NULL",
        )
    } else {
        sqlx::query_scalar::<_, i64>(
            "SELECT COUNT(*) FROM operators WHERE role = 'node_operator'",
        )
    }
    .fetch_one(pool)
    .await
    .unwrap_or(0)
}

async fn count_witness_origins(pool: &PgPool) -> i64 {
    if !column_exists(pool, "witness_observations", "origin").await {
        return 0;
    }
    sqlx::query_scalar::<_, i64>(
        "SELECT COUNT(DISTINCT origin) FROM witness_observations",
    )
    .fetch_one(pool)
    .await
    .unwrap_or(0)
}

async fn count_issued_sbts(pool: &PgPool) -> i64 {
    let has_revoked = column_exists(pool, "key_credentials", "revoked_at").await;
    let has_sbt = column_exists(pool, "key_credentials", "sbt_nontransferable").await;

    // If neither column exists, the table itself may not exist — return 0.
    if !has_revoked && !has_sbt {
        let table_exists = column_exists(pool, "key_credentials", "id").await
            || sqlx::query_scalar::<_, bool>(
                "SELECT EXISTS(SELECT 1 FROM information_schema.tables \
                 WHERE table_schema = 'public' AND table_name = 'key_credentials')",
            )
            .fetch_one(pool)
            .await
            .unwrap_or(false);
        if !table_exists {
            return 0;
        }
    }

    let query = match (has_revoked, has_sbt) {
        (true, true) => "SELECT COUNT(*) FROM key_credentials \
                         WHERE revoked_at IS NULL AND sbt_nontransferable = true",
        (true, false) => "SELECT COUNT(*) FROM key_credentials WHERE revoked_at IS NULL",
        (false, true) => "SELECT COUNT(*) FROM key_credentials WHERE sbt_nontransferable = true",
        (false, false) => "SELECT COUNT(*) FROM key_credentials",
    };
    sqlx::query_scalar::<_, i64>(query)
        .fetch_one(pool)
        .await
        .unwrap_or(0)
}

async fn count_distinct_shards(pool: &PgPool) -> i64 {
    // Collect distinct shard_id values across all relevant tables via a UNION.
    // Each branch is guarded by an information_schema existence check so the
    // query degrades gracefully when tables don't exist yet.
    let tables = [
        "ingestion_proofs",
        "doc_commits",
        "dataset_artifacts",
        "dataset_lineage_events",
    ];
    let mut parts: Vec<&str> = Vec::new();
    for t in &tables {
        if column_exists(pool, t, "shard_id").await {
            parts.push(t);
        }
    }
    if parts.is_empty() {
        return 0;
    }
    // Build a safe UNION query — table names are validated against the
    // compile-time allowlist above, never from user input.
    let union_sql = parts
        .iter()
        .map(|t| format!("SELECT shard_id FROM \"{}\" WHERE shard_id IS NOT NULL", t))
        .collect::<Vec<_>>()
        .join(" UNION ");
    let sql = format!("SELECT COUNT(*) FROM ({}) AS _shards", union_sql);
    sqlx::query_scalar::<_, i64>(&sql)
        .fetch_one(pool)
        .await
        .unwrap_or(0)
}

async fn count_public_proofs(pool: &PgPool) -> i64 {
    let mut total: i64 = 0;

    // ingestion_proofs: count all rows
    let ingestion_exists: bool = sqlx::query_scalar::<_, bool>(
        "SELECT EXISTS(SELECT 1 FROM information_schema.tables \
         WHERE table_schema = 'public' AND table_name = 'ingestion_proofs')",
    )
    .fetch_one(pool)
    .await
    .unwrap_or(false);
    if ingestion_exists {
        total += sqlx::query_scalar::<_, i64>("SELECT COUNT(*) FROM ingestion_proofs")
            .fetch_one(pool)
            .await
            .unwrap_or(0);
    }

    // Non-empty proof columns in other tables
    for (table, col) in [
        ("doc_commits", "zk_proof"),
        ("dataset_artifacts", "zk_proof"),
        ("credential_ledger_events", "inclusion_proof"),
    ] {
        if column_exists(pool, table, col).await {
            let sql = format!(
                "SELECT COUNT(*) FROM \"{}\" WHERE {} IS NOT NULL AND {} != ''",
                table, col, col
            );
            total += sqlx::query_scalar::<_, i64>(&sql)
                .fetch_one(pool)
                .await
                .unwrap_or(0);
        }
    }
    total
}

pub fn format_uptime(seconds: u64) -> String {
    if seconds < 60 {
        format!("{}s", seconds)
    } else if seconds < 3600 {
        format!("{}m", seconds / 60)
    } else if seconds < 86400 {
        format!("{}h", seconds / 3600)
    } else {
        format!("{}d", seconds / 86400)
    }
}

// ─── tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format_uptime_seconds() {
        assert_eq!(format_uptime(0), "0s");
        assert_eq!(format_uptime(59), "59s");
    }

    #[test]
    fn format_uptime_minutes() {
        assert_eq!(format_uptime(60), "1m");
        assert_eq!(format_uptime(3599), "59m");
    }

    #[test]
    fn format_uptime_hours() {
        assert_eq!(format_uptime(3600), "1h");
        assert_eq!(format_uptime(86399), "23h");
    }

    #[test]
    fn format_uptime_days() {
        assert_eq!(format_uptime(86400), "1d");
        assert_eq!(format_uptime(172800), "2d");
    }
}
